lxd/device/disk.go

```diff
@@ -431,34 +431,92 @@ func (d *disk) validateConfig(instConf instance.ConfigReader) error {
 			return fmt.Errorf("Storage volumes cannot be specified as absolute paths")
 		}
 
-		// Only perform expensive instance pool volume checks when not validating a profile and after
-		// device expansion has occurred (to avoid doing it twice during instance load).
-		if d.inst != nil && !d.inst.IsSnapshot() && len(instConf.ExpandedDevices()) > 0 {
+		var dbVolume *db.StorageVolume
+		var storageProjectName string
+
+		if d.inst != nil && !d.inst.IsSnapshot() && d.config["source"] != "" && d.config["path"] != "/" {
 			d.pool, err = storagePools.LoadByName(d.state, d.config["pool"])
 			if err != nil {
 				return fmt.Errorf("Failed to get storage pool %q: %w", d.config["pool"], err)
 			}
 
+			// Derive the effective storage project name from the instance config's project.
```

Comment:
This section seems too complicated and is quite hard to follow.

Please can you break out the actual check logic into a separate function on the `disk` struct and then call it from here so we can more easily reason about where its call from.

Reply:
I pushed a new version that simplifies the conditions and makes the `validateConfig` easier to read and follow. That may also help you evaluate the initialization of `d.pool`, as discussed on the previous conversation. If you think this changeis less readable or complicates future changes let me know. Also I can add the function you requested if it is still necessary.